### DISCUSSION TITLE
Client-side window decorations: do them in the character grid like the tab bar

### DISCUSSION THREAD

**Discussion by valpackett** [wezterm/wezterm#1125] [Ideas]
Just had a really big brain idea: what if client-side decorations were rendered just like the tab bar — using terminal characters?

- no need for SCTK frame code anymore on Wayland
- users on other platforms could opt in to them if they like the look!
  - and that could make debugging a bit easier
- this would imply making the titlebar line a draggable area (i.e. it would drag the window) so it would be easy to also make the free space of the tab bar drag the window as well!
- fun option: render it on the bottom just like the tab bar

**Comment by wez** [comment]
I really hate that the current SCTK based frame stuff is out of sync with the rest of the Window; it feels so janky and behavior varies across the compositors that I've tried.  It's maddening!

I'm very slowly (due to availble personal bandwidth!) advancing towards something sort of like what you described.  Today, you can turn off decorations on most platforms and still resize the window by the edges (assuming RESIZE is still enabled) and drag by the tab bar, so the concept is basically proven.

Wayland is a bit problematic because it seems to want to make it really awkward to do direct interactive resizes/moves, so dragging the window is essentially a request to ask the compositor to initiate a drag for the user and then the app is out of the loop.  I don't recall if the tab bar dragging currently works on wayland or whether it is flakey.

Ignoring that(!) for the moment, in broad strokes what's on my mind is:

* Introduce an alternative rendering of the tab bar so that it is similar to the integrated tab/title bar seen in Chrome and Firefox  including the ability to use proportional fonts for tab titles (better information density).
* as a result, that rendering would be distinct from the current terminal matrix - the reason the tab bar is in the matrix today was laziness/convenience.  Today we have a much more flexible Quad allocator that allows for more dynamically produced content outside of the (originally) statically allocated terminal grid, and with ab8c8dc6c9877dc02a270f59103f2be138697d18 and a7b20443bafd520276177b433e7e2f3583391842 its a bit easier to add more UI elements that interact with the mouse
* Figure out how to reconcile the existing cell-based tab/status area formatting events that a number of users have customized to deal with proportional fonts.
* I hadn't planned to do anything in particular about resize decorations, as they're generally invisible these days anyway
* With the new tab bar rendering, implement dragging to re-order tabs (#549)

**Reply by valpackett** [reply]
> I don't recall if the tab bar dragging currently works on wayland or whether it is flakey.

Just doesn't work. I didn't think it was a thing already, I'll investigate why it doesn't work! (upd:  #1127)

> the reason the tab bar is in the matrix today was laziness/convenience

Of course, but I think it also works out as an interesting aesthetic choice.
